### PR TITLE
solution 24/12

### PR DESCRIPTION
**Pull request by VadimKadasiev**

### CODE REVIEW

File: src/main/java/core/basesyntax/Main.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    private static final String PURCHASE_OPERATION_CODE = "p";
+    private static final String SUPPLY_OPERATION_CODE = "s";
+    private static final String RETURN_OPERATION_CODE = "r";
+    private static final Map<String, OperationHandler> operationHandlerMap = new HashMap<>();
+
+    public static void main(String[] args) {
+        operationHandlerMap.put(BALANCE_OPERATION_CODE, new BalanceHandler());
+        operationHandlerMap.put(PURCHASE_OPERATION_CODE, new PurchaseHandler());
+        operationHandlerMap.put(RETURN_OPERATION_CODE, new AddFruitQuantity());
+        operationHandlerMap.put(SUPPLY_OPERATION_CODE, new AddFruitQuantity());
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/Main.java:32]: You can init by Map.of

File: src/main/resources/input.csv [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (Sanderfuck) [inline src/main/resources/input.csv:25]: For what a lot of empty lines?

File: src/main/java/core/basesyntax/db/FruitStorage.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package core.basesyntax.db;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class FruitStorage {
+
+    public static final Map<String, Integer> storageData = new HashMap<>();
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/db/FruitStorage.java:8]: image

File: src/main/java/core/basesyntax/db/InputFileHandler.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package core.basesyntax.db;
+
+import java.util.List;
+import java.util.Scanner;
+
+public interface InputFileHandler {
+
+    List<String> inputData(Scanner scanner);
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/db/InputFileHandler.java:8]: image

File: src/main/java/core/basesyntax/Main.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.Map;
+import java.util.Scanner;
+
+public class Main {
+    private static final String INPUT_FILE_PATH = "jv-fruit-shop/src/main/resources/input.csv";
+    private static final String REPORT_FILE_PATH = "jv-fruit-shop/src/main/resources/report.csv";
+    private static final String BALANCE_OPERATION_CODE = "b";
+    private static final String PURCHASE_OPERATION_CODE = "p";
+    private static final String SUPPLY_OPERATION_CODE = "s";
+    private static final String RETURN_OPERATION_CODE = "r";
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/Main.java:25]: Create enum for this

File: src/main/java/core/basesyntax/db/ReportFileWriterImpl.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+public class ReportFileWriterImpl implements ReportFileWriter {
+    private static final String SEPARATOR = ",";
+
+    @Override
+    public void writeToFile(Map<String, Integer> fruitsStorage, String filePath) {
+        FileWriter writer;
+        try {
+            writer = new FileWriter(filePath);
+            writer.write("fruit,quantity" + System.lineSeparator());
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/db/ReportFileWriterImpl.java:15]: Save String in constant

File: src/main/java/core/basesyntax/strategy/OperationStrategyImpl.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    public OperationStrategyImpl(Map<String, OperationHandler> operationHandlerMap) {
+        this.operationHandlerMap = operationHandlerMap;
+    }
+
+    @Override
+    public void operationPerform(List<String> inputData) {
+        for (String dataLine : inputData) {
+            try {
+                String[] splitDataLine = dataLine.split(SEPARATOR);
+                if (!Objects.equals(splitDataLine[0], "fruit")) {
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/strategy/OperationStrategyImpl.java:24]: Magic number and string

**Review by Sanderfuck — CHANGES_REQUESTED** [review]
Before start write code, read readme and checklist file

File: src/main/java/core/basesyntax/handlers/AddFruitQuantity.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package core.basesyntax.handlers;
+
+import core.basesyntax.db.FruitStorage;
+
+public class AddFruitQuantity implements OperationHandler {
```
Comment (olekskov) [inline src/main/java/core/basesyntax/handlers/AddFruitQuantity.java:5]: 1) Verbs are used for methods naming
2) Strange to see BalanceHandler, PurchaseHandler and AddFruitQuantity as equal implemetations of single interface
```suggestion
public class QuantityAdditionHandler implements OperationHandler {

```

File: src/main/java/core/basesyntax/strategy/OperationStrategyImpl.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    @Override
+    public void operationPerform(List<String> inputData) {
+        for (String dataLine : inputData) {
+            try {
+                String[] splitDataLine = dataLine.split(SEPARATOR);
+                if (!Objects.equals(splitDataLine[OPERATION_INDEX], HEADER_FIRST_VALUE)) {
+                    operationHandlerMap.get(splitDataLine[OPERATION_INDEX])
+                            .calculateResult(splitDataLine[FRUIT_INDEX],
+                                    Integer.parseInt(splitDataLine[QUANTITY_INDEX]));
+                }
```
Comment (olekskov) [inline src/main/java/core/basesyntax/strategy/OperationStrategyImpl.java:29]: barely readable code
Please introduce some variable like
String fruitName = splitDataLine[FRUIT_INDEX];
int fruitQuantity = Integer.parseInt(splitDataLine[QUANTITY_INDEX]);
and may be some others to make this part understandable without deepdiving into it

File: src/main/java/core/basesyntax/handlers/BalanceHandler.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+import core.basesyntax.db.FruitStorage;
+
+public class BalanceHandler implements OperationHandler {
+    @Override
+    public void calculateResult(String fruit, int quantity) {
+        if (quantity < 0) {
+            throw new RuntimeException("Balance value can't be negative");
+        } else {
+            FruitStorage.storageData.put(fruit,quantity);
+        }
```
Comment (olekskov) [inline src/main/java/core/basesyntax/handlers/BalanceHandler.java:12]: We do not need else block in the end of method, it doesn't produce any difference in logic
```suggestion
    public void calculateResult(String fruit, int quantity) {
        if (quantity < 0) {
            throw new RuntimeException("Balance value can't be negative");
        }
        FruitStorage.storageData.put(fruit,quantity);
```
Please recheck similar cases

**Review by olekskov — CHANGES_REQUESTED** [review]
Seems good overall but requires code-style changes

File: src/main/java/core/basesyntax/handlers/QuantityAdditionHandler.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    public void calculateResult(String fruit, int quantity) {
+        if (FruitStorage.storageData.get(fruit) != null) {
+            FruitStorage.storageData.put(fruit,
+                    FruitStorage.storageData.get(fruit) + quantity);
+        } else {
+            FruitStorage.storageData.put(fruit, quantity);
+        }
+    }
+}
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/handlers/QuantityAdditionHandler.java:15]:
```suggestion

```

File: src/main/java/core/basesyntax/db/InputFileReaderImpl.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+        Scanner scanner;
+        try {
+            scanner = new Scanner(new File(filePath));
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException("Can't find the input file", e);
+        }
+        return scanner;
+    }
+}
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/db/InputFileReaderImpl.java:19]:
```suggestion
```

File: src/main/java/core/basesyntax/db/ReportFileWriterImpl.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+                            + fruitsStorage.get(fruitName) + System.lineSeparator());
+                }
+            }
+            writer.close();
+        } catch (IOException e) {
+            throw new RuntimeException("Can`t write data to file", e);
+        }
+    }
+}
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/db/ReportFileWriterImpl.java:32]: each file should end with exactly one empty line

File: src/main/java/core/basesyntax/db/ReportFileWriterImpl.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+                    throw new RuntimeException("Quantity value can't be negative");
+                }
+                if (fruitsStorage.get(fruitName) > 0) {
+                    writer.write(fruitName + SEPARATOR
+                            + fruitsStorage.get(fruitName) + System.lineSeparator());
+                }
+            }
+            writer.close();
+        } catch (IOException e) {
+            throw new RuntimeException("Can`t write data to file", e);
```
Comment (okuzan) [inline src/main/java/core/basesyntax/db/ReportFileWriterImpl.java:28]: always include filename in exception messages related to files

File: src/main/java/core/basesyntax/db/ReportFileWriterImpl.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.Map;
+
+public class ReportFileWriterImpl implements ReportFileWriter {
+    private static final String SEPARATOR = ",";
+    private static final String HEADER = "fruit,quantity";
+
+    @Override
+    public void writeToFile(Map<String, Integer> fruitsStorage, String filePath) {
+        FileWriter writer;
+        try {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/db/ReportFileWriterImpl.java:14]: use try with resources

File: src/main/java/core/basesyntax/handlers/BalanceHandler.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
+package core.basesyntax.handlers;
+
+import core.basesyntax.db.FruitStorage;
+
+public class BalanceHandler implements OperationHandler {
+    @Override
+    public void calculateResult(String fruit, int quantity) {
+        if (quantity < 0) {
+            throw new RuntimeException("Balance value can't be negative");
```
Comment (okuzan) [inline src/main/java/core/basesyntax/handlers/BalanceHandler.java:9]: include quantity in the message

File: src/main/java/core/basesyntax/strategy/OperationStrategyImpl.java [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+            String operationType = splitDataLine[OPERATION_INDEX];
+            try {
+                if (!Objects.equals(operationType, HEADER_FIRST_VALUE)) {
+                    String fruitType = splitDataLine[FRUIT_INDEX];
+                    int quantity = Integer.parseInt(splitDataLine[QUANTITY_INDEX]);
+                    operationHandlerMap.get(operationType)
+                            .calculateResult(fruitType, quantity);
+                }
+            } catch (RuntimeException e) {
+                throw new RuntimeException("Wrong data format", e);
```
Comment (okuzan) [inline src/main/java/core/basesyntax/strategy/OperationStrategyImpl.java:33]: include dataLine

File: src/main/java/core/basesyntax/Main.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+import core.basesyntax.handlers.QuantityAdditionHandler;
+import core.basesyntax.strategy.OperationStrategy;
+import core.basesyntax.strategy.OperationStrategyImpl;
+import java.util.List;
+import java.util.Map;
+import java.util.Scanner;
+
+public class Main {
+    private static final String INPUT_FILE_PATH = "jv-fruit-shop/src/main/resources/input.csv";
+    private static final String REPORT_FILE_PATH = "jv-fruit-shop/src/main/resources/report.csv";
```
Comment (okuzan) [inline src/main/java/core/basesyntax/Main.java:20]:
```suggestion
    private static final String INPUT_FILE_PATH = "src/main/resources/input.csv";
    private static final String REPORT_FILE_PATH = "src/main/resources/report.csv";
```
will this work?

File: src/main/java/core/basesyntax/db/InputFileReaderImpl.java [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.Scanner;
+
+public class InputFileReaderImpl implements InputFileReader {
+    @Override
+    public Scanner readFromFile(String filePath) {
+        Scanner scanner;
+        try {
+            scanner = new Scanner(new File(filePath));
```
Comment (okuzan) [inline src/main/java/core/basesyntax/db/InputFileReaderImpl.java:12]: IMO a bit weird to just return a scanner, return the content , combine this with inputfilehandler

File: src/main/java/core/basesyntax/db/InputFileReaderImpl.java [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException("Can't find the input file", e);
```
Comment (okuzan) [inline src/main/java/core/basesyntax/db/InputFileReaderImpl.java:14]: + filename

**Review by okuzan — CHANGES_REQUESTED** [review]
Code style issues and a few suggestions

File: src/main/java/core/basesyntax/Main.java (hunk 18 above)
Comment (VadimKadasiev) [reply src/main/java/core/basesyntax/Main.java:20]: I have no idea, why it doesn't work in my IDE...

File: src/main/java/core/basesyntax/db/InputFileReaderImpl.java (hunk 19 above)
Comment (VadimKadasiev) [reply src/main/java/core/basesyntax/db/InputFileReaderImpl.java:12]: OK. I thought, that it's a task condition to split these two processes.

File: src/main/java/core/basesyntax/db/InputFileHandlerImpl.java [hunk 21]
```diff
@@ -0,0 +1,25 @@
+package core.basesyntax.db;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+public class InputFileHandlerImpl implements InputFileHandler {
+    private final List<String> inputList = new ArrayList<>();
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/db/InputFileHandlerImpl.java:10]: You don`t need this field, use this variable inside method

File: src/main/java/core/basesyntax/handlers/QuantityAdditionHandler.java [hunk 22]
```diff
@@ -0,0 +1,14 @@
+package core.basesyntax.handlers;
+
+import core.basesyntax.db.FruitStorage;
+
+public class QuantityAdditionHandler implements OperationHandler {
+    public void calculateResult(String fruit, int quantity) {
+        if (FruitStorage.storageData.get(fruit) != null) {
+            FruitStorage.storageData.put(fruit,
+                    FruitStorage.storageData.get(fruit) + quantity);
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/handlers/QuantityAdditionHandler.java:9]: You can use merge values method of map instead of this check

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

**Review by olekskov — APPROVED** [review]

**Pull request closed without merging**